File: src/Homework1/Homework1/Parser.cs
using System;

namespace Homework1
{
    public static class Parser
    {
        private const int WrongArgsNumber = 1;
        private const int WrongValueFormat = 2;
        private const int WrongOperation = 3;
        private const int DivisionByZero = 4;

        public static int ParseArgs(
            string[] args,
            out int val1,
            out Calculator.SupportedOperations operation,
            out int val2)
        {
            val1 = val2 = (int)(operation = 0);
            if (Parser.CheckArgsNumber(args)) return WrongArgsNumber;
            if (Parser.TryParseValue(args[0], out val1)
                || Parser.TryParseValue(args[2], out val2)) return WrongValueFormat;
            if (Parser.TryParseOperator(args[1], out operation)) return WrongOperation;
            if (val2 == 0) return DivisionByZero;
            return 0;
        }

        private static bool CheckArgsNumber(string[] args)
        {
            if (args.Length == 3) return false;
            Console.WriteLine($"Programm needs 3 args, but there is {args.Length}");
            return true;
        }

        private static bool TryParseValue(string arg, out int result)
        {
            if (int.TryParse(arg, out result))
                return false;
            Console.WriteLine($"{arg} is not int");
            return true;
        }

        private static bool TryParseOperator(string arg, out Calculator.SupportedOperations operation)
        {
            switch (arg)
            {
                case "+":
                    operation = Calculator.SupportedOperations.Plus;
                    break;
                case "-":
                    operation = Calculator.SupportedOperations.Minus;
                    break;
                case "*":
                    operation = Calculator.SupportedOperations.Multiply;
                    break;
                case "/":
                    operation = Calculator.SupportedOperations.Divide;
                    break;
                default:
                    operation = default;
                    return true;
            }

            return false;
        }
    }
}

File: src/Homework4/Homework4/Parser.fs
module Homework4.Parser

open System

type Parser =
    | Correct = 0
    | WrongArgsNumber = 1
    | WrongValueFormat = 2
    | WrongOperation = 3
    | DivisionByZero = 4

let checkArgsNumber (args: string []) =
    if args.Length = 3 then
        true
    else
        printf $"Program needs 3 args, but there is {args.Length}"
        false

let tryParseOperator (arg: string) (operation: outref<Calculator.SupportedOperations>) =
    operation <- match arg with
    | "+" -> Calculator.SupportedOperations.Plus
    | "-" -> Calculator.SupportedOperations.Minus
    | "/" -> Calculator.SupportedOperations.Divide
    | _ -> Calculator.SupportedOperations.Multiply
    operation <> Calculator.SupportedOperations.Multiply || arg = "*"

let parseArgs
    (args: string [])
    (val1: outref<int>)
    (operation: outref<Calculator.SupportedOperations>)
    (val2: outref<int>)
    =
    if not (checkArgsNumber args) then
        Parser.WrongArgsNumber
    elif not (Int32.TryParse(args.[0], &val1))
         || not (Int32.TryParse(args.[2], &val2)) then
        Parser.WrongValueFormat
    elif not (tryParseOperator args.[1] &operation) then
        Parser.WrongOperation
    elif (val2 = 0) && (args.[1]="/") then
        Parser.DivisionByZero
    else
        Parser.Correct
